src/Nethermind/Nethermind.Network/TrustedNodes/TrustedNodesManager.cs

```diff
@@ -0,0 +1,194 @@
+
+namespace Nethermind.Network
+{
+    public class TrustedNodesManager : ITrustedNodesManager
+    {
+        private ConcurrentDictionary<PublicKey, NetworkNode> _nodes = new();
+        private readonly string _trustedNodesPath;
+        private readonly ILogger _logger;
+        private readonly Channel<Node> _nodeChannel = Channel.CreateBounded<Node>(
+        new BoundedChannelOptions(1 << 16)  // capacity of 2^16 = 65536
+        {
+            // "Wait" to have writers wait until there is space.
+            FullMode = BoundedChannelFullMode.Wait
+        });
+
+        public TrustedNodesManager(string trustedNodesPath, ILogManager logManager)
+        {
+            _trustedNodesPath = trustedNodesPath.GetApplicationResourcePath();
+            _logger = logManager.GetClassLogger();
+        }
+
+        public IEnumerable<NetworkNode> Nodes => _nodes.Values;
+
+        public async Task InitAsync()
+        {
+            if (!File.Exists(_trustedNodesPath))
+            {
+                if (_logger.IsDebug) _logger.Debug($"Trusted nodes file not found at: {_trustedNodesPath}");
+            }
+
+            var nodes = new ConcurrentDictionary<PublicKey, NetworkNode>();
+
+            await foreach (string line in File.ReadLinesAsync(_trustedNodesPath))
+            {
+                if (string.IsNullOrWhiteSpace(line))
+                {
+                    continue;
+                }
+                try
+                {
+                    NetworkNode node = new NetworkNode(line);
+                    nodes.TryAdd(node.NodeId, node);
+                }
+                catch (Exception ex) when (ex is ArgumentException or SocketException)
+                {
+                    if (_logger.IsError)
+                    {
+                        _logger.Error($"Failed to parse '{line}' as a trusted node.", ex);
+                    }
+                }
+            }
+
+            if (_logger.IsInfo)
+            {
+                _logger.Info($"Loaded {nodes.Count} trusted nodes from: {Path.GetFullPath(_trustedNodesPath)}");
+            }
+            if (_logger.IsDebug && nodes.Any())
+            {
+                _logger.Debug("Trusted nodes:\n" + string.Join(Environment.NewLine, nodes.Values.Select(n => n.ToString())));
+            }
+
+            _nodes = nodes;
+        }
+
+
+        // ---- INodeSource requirement: IAsyncEnumerable<Node> ----
+        public async IAsyncEnumerable<Node> DiscoverNodes([EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            // yield existing nodes.
+            foreach (NetworkNode netNode in _nodes.Values)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                yield return new Node(netNode) { IsTrusted = true };
+            }
+
+            // continuously yield new nodes as they are added via the channel.
+            while (await _nodeChannel.Reader.WaitToReadAsync(cancellationToken))
```

Comment:
Instead of waiting and doing two `while`s, could you just `_nodeChannel.Reader.ReadAllAsync`